034-kafka-topic-replication-factor.md

```diff
@@ -0,0 +1,76 @@
+
+## Current situation
+
+Currently, users have two options to change the topic RF:
+* Leverage the Native Kafka script to change the RF. Users need to generate the partition replica assignment and balance the kafka clusters manually if needed. 
+* Delete the topic and recreate the topic with the desired RF.
+
+Clearly none of the above approaches is user-friendly.
+
+When users attempt to change the RF in the KafkaTopic Resources, the Status of the KafkaTopic will become `NotReady` with `ReplicationFactorChangeException`. The following error message is shared with users.
+
+`Changing 'spec.replicas' is not supported. This KafkaTopic's 'spec.replicas' should be reverted to 4 and then the replication should be changed directly in Kafka.`
+
+## Motivation
+
+With this new feature, it will become much easier to achieve the desired tradeoff between the cost and data reliability in the kafka clusters. 
+
+For instance, users can lower the RFs for selective less critical topics to reduce the infrastructure cost (Compute, Storage and Network). Also, for some critical kafka topics, users can increase the data reliability through the bump of the RFs ONLY for these specific topics. 
+
+More importantly, users only need to specify the new RF number in the `KafkaTopic` resource and approve the change, in order to change the RF for kafka topics. The heavy lifting work, such as new replica placement, extra replica deletion and possible cluster rebalance, is completely taken care by the strimzi. 
+
+## Proposal
+
+### API
+The Kafka Topic RF can be changed by users in the following steps:
+* Update the `spec.replicas` properties in the target `KafkaTopic` Resource in the kubernetes cluster
+* Watch the `Status` field of the `KafkaTopic Resource`, for the status update of this RF change.
+
+### Design
+We propose to change the topic RFs with the support of `Cruise Control` and `Strimzi Topic Operator`. 
+
+#### Cruise Control
+Cruise Control Service will be leveraged to change the RF for topics in the kafka clusters, for the following reasons: 
+* Strimzi supports the Cruise Control Service, and it's ready for use with simple configurations. 
+* Cruise Control has native support for the RF changes. It automates the whole process, including selecting of new replicas and extra replicas, triggering the replica creation in the kafka brokers and monitoring the cluster during the process.
+
+#### Kafka Topic Operator
+Strimzi Topic Operator will watch the changes of `KafkaTopic` Resource. 
+
+For any new topic RF change event, the `Strimzi Topic Operator` will validate the change. For instance, the new RF should be >= the min insync replica set size for the target topic. 
+
+For the valid changes, the `Strimzi Topic Operator` will send requests to `Cruise Control`. The Kafka Topic Operator will periodically poll the status from the Cruise Control, and update the status in the `Status` field of the `KafkaTopic` resource.
```

Comment:
I don't know the CC API, but does it return back some kind of cookie or identifier for the request that was made to its endpoint and that needs to be passed when polling for completion? If so, I assume this needs to be added to the status section of the CR. An example CR would help.

We should also consider the various erroneous flows:

1. What should the operator do if the topic is deleted (and possibly recreated) while a change in RF is happening?
2. What happens if CC gets restarted while a change in RF is happening (I assume CC has persisted its state somewhere, so this isn't a problem, but if that's the case let's say so).
3. What happens in the user changes the RF again, while a change in RF already is happening? Does the TO have to cancel the existing change, or wait for it to complete?
4. What happens if the TO crashes after make the request to CC to change the RF but before it's able to update the `KafkaTopic.status` (assuming we are storing some kind of request id/cookie there)?

Reply:
As I mentioned in a previous comment, I think we would need a state machine based on the interaction with the CC endpoint. I would try it and report how it works in the proposal, by describing a potential state machine in the TO. I guess it would need some user interaction to approve, unless we want the TO doing it automatically.

Reply:
AFAIK the flow would be almost identical to the process and logic used for a full cluster rebalance, but initiated by the TO. I wonder if it makes sense to have a common way to execute these CC (POST) interactions, otherwise we are likely to be replicating a lot of existing logic from the KRAO in the TO?